### ISSUE TITLE
Panic in Kubernetes when using DSN from Secret

### ISSUE THREAD

**Issue by hysmio** [ory/hydra#1545]
**Describe the bug**

When attempting to use a Kubernetes (k8s) Secret to store the DSN, the container crashes almost immediately with an **invalid memory address or nil pointer dereference**

This only occurs when using the a k8s Secret, I am able to successfully use hydra in k8s when setting the value explicitly.

**Reproducing the bug**

1. Run `kubectl apply -f hydra.yaml` (hydra.yaml being the manifest above)
2. Deployment tries to start a Pod, but it will die almost immediately.

*Server logs*
Calling `kubectl logs` on a dead Pod gives this
```Config file not found because "Config File ".hydra" Not Found in "[/]""
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xd8 pc=0xac53f9]

goroutine 1 [running]:
github.com/ory/x/urlx.ParseOrFatal(0x0, 0x0, 0xc000036084, 0x39, 0x0)
	/go/pkg/mod/github.com/ory/x@v0.0.64/urlx/parse.go:22 +0x69
github.com/ory/hydra/driver.(*RegistrySQL).CanHandle(0xc000273ef0, 0xc000036084, 0x39, 0x0)
	/go/src/github.com/ory/hydra/driver/registry_sql.go:160 +0x54
github.com/ory/x/dbal.GetDriverFor(0xc000036084, 0x39, 0x39, 0xc00066bb80, 0xac5738, 0xc00016a1e0)
	/go/pkg/mod/github.com/ory/x@v0.0.64/dbal/driver.go:36 +0x85
github.com/ory/hydra/driver.NewRegistry(0x105dec0, 0xc00016a1e0, 0x105dec0, 0xc00016a1e0, 0xc00066bc18, 0x86b021)
	/go/src/github.com/ory/hydra/driver/registry.go:59 +0x61
github.com/ory/hydra/driver.NewDefaultDriver(0x1058580, 0xc00016a000, 0x0, 0x1770720, 0x0, 0x0, 0x1011100, 0xf, 0x1040f80, 0x28, ...)
	/go/src/github.com/ory/hydra/driver/driver_default.go:21 +0xf6
github.com/ory/hydra/cmd/server.RunServeAll.func1(0x173c140, 0x1770720, 0x0, 0x0)
	/go/src/github.com/ory/hydra/cmd/server/handler.go:130 +0x159
github.com/spf13/cobra.(*Command).execute(0x173c140, 0x1770720, 0x0, 0x0, 0x173c140, 0x1770720)
	/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:766 +0x2ae
github.com/spf13/cobra.(*Command).ExecuteC(0x173ba20, 0xdba640, 0x2b, 0x2c)
	/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852 +0x2ec
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
github.com/ory/hydra/cmd.Execute()
	/go/src/github.com/ory/hydra/cmd/root.go:59 +0x32
main.main()
	/go/src/github.com/ory/hydra/main.go:33 +0x50
```

*Server configuration*

Kubernetes manifest file
```yaml
# Create Hydra service initially
apiVersion: v1
kind: Service
metadata:
  name: hydra
  labels:
    app: hydra
spec:
  selector:
    app: hydra
  ports:
    - port: 4444
      name: hydra-public
      # targetPort: 4444 # Not neccessary -> By default and for convenience, the targetPort is set to the same value as the port field.
    - port: 4445
      name: hydra-admin
      # targetPort: 4445
---
# Deploy the Hydra instance
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hydra-deployment
spec:
  selector:
    matchLabels:
      app: hydra
  replicas: 1
  template:
    metadata:
      labels:
        app: hydra
    spec:
      containers:
        - name: hydra-container
          image: oryd/hydra:latest
          imagePullPolicy: Always
          env:
            - name: DSN
              valueFrom:
                secretKeyRef:
                  name: hydra-secrets
                  key: DSN
            - name: SECRETS_SYSTEM
              valueFrom:
                secretKeyRef:
                  name: hydra-secrets
                  key: SECRETS_SYSTEM
            - name: URLS_SELF_ISSUER
              value: **omitted**
            - name: URLS_CONSENT
              value: **omitted**
            - name: URLS_LOGIN
              value: **omitted**
          ports:
            - containerPort: 4444
            - containerPort: 4445
```

**Environment**

* Version: oryd/hydra:v1.0.0
* Environment: Kubernetes, Google Kubernetes Engine, Cloud SQL Postgres

**Comment by aeneasr** [comment]
This is most likely caused by a malformed or missing DSN value. Unfortunately, we’re not able to properly validate env vars as of now, this will be solved in a future update.

However this should obviously not panic

**Comment by hysmio** [comment]
@aeneasr 

> This is most likely caused by a malformed or missing DSN value.

I used the exact same secret in an Ubuntu image to confirm the DSN, the secret was decoded successfully.

The DSN follows the syntax: `postgres://hydra:{alpha numeric string, no symbols}@{gcloud host}/hydra`

When given that same DSN by simply using `value` instead of `valueFrom`, Hydra works successfully, though it sounds like a k8s issue, I thought I should open an issue here to let you know.

**Comment by aeneasr** [comment]
I'm pretty sure it's a configuration issue - chack our helm chart implementation for reference ( https://github.com/ory/k8s/tree/master/helm/charts/hydra ) - or, even better, use it!

**Comment by hysmio** [comment]
I'll attempt to use it tomorrow and let you know my results! Thanks!

**Comment by aeneasr** [comment]
I just confirmed that this is fixed on master. I also confirmed that this can only be caused by a missing/empty DSN string, so your Kubernetes config is probably broken.

I'm closing this, the fix of the panic will be released with 1.0.1

**Comment by hysmio** [comment]
I've investigated what the issue was in my configuration and actually found that when I was using:
```bash
echo -e "postgres://some-dsn" | base64
```
It was appending a new line or carriage return & new line, which I believe caused the crash, I've redone the secrets using vscode's base64 conversion & have been much more successful.

TLDR: Crashes when DSN includes newline

**Comment by haf** [comment]
Same experience as hysmio, caused by overeager IDEs adding newlines to secret files... The above is a good repro; @aeneasr could consider trimming the string or at least warning.

**Comment by aeneasr** [comment]
As already pointed out in my previous comment https://github.com/ory/hydra/issues/1545#issuecomment-526116942

> I'm closing this, the fix of the panic will be released with 1.0.1

This is already resolved. So please either include new information ("this does not work on 1.0.8") relevant for this issue or upgrade.
